dmoj/cptbox/helper.cpp

```diff
@@ -90,13 +91,62 @@ int cptbox_child_run(const struct child_config *config) {
     kill(getpid(), SIGSTOP);
 
 #if !PTBOX_FREEBSD
+    // landlock setup
+    int ruleset_fd, rc;
+    struct landlock_ruleset_attr ruleset_attr = {
+        .handled_access_fs = LANDLOCK_ACCESS_FS_EXECUTE | LANDLOCK_ACCESS_FS_WRITE_FILE | LANDLOCK_ACCESS_FS_READ_FILE |
+                             LANDLOCK_ACCESS_FS_READ_DIR | LANDLOCK_ACCESS_FS_REMOVE_DIR |
+                             LANDLOCK_ACCESS_FS_REMOVE_FILE | LANDLOCK_ACCESS_FS_MAKE_REG |
+                             LANDLOCK_ACCESS_FS_MAKE_SYM | LANDLOCK_ACCESS_FS_MAKE_DIR | LANDLOCK_ACCESS_FS_REFER,
+    };
+
+    int landlock_version = get_landlock_version();
+    if (landlock_version < 2) {
+        // ABI not old enough. Skip to seccomp
+        goto seccomp_setup;
+    }
+
+    ruleset_fd = landlock_create_ruleset(&ruleset_attr, sizeof(ruleset_attr), 0);
+    if (ruleset_fd < 0) {
+        perror("Failed to create a ruleset");
+        return PTBOX_SPAWN_FAIL_LANDLOCK;
+    }
+
+    // Note: WRITE must imply READ. This is required because even if one rule allows writing and another allows reading,
```

Comment:
The `O_TRUNC` handling in Landlock scares me. Could we add some testcases with a custom grader that make sure that we don't allow truncating etc.?